socks5.py

```diff
@@ -220,57 +376,131 @@ def read_addrport(self, address_type, sockfile):
         port, = readstruct(sockfile, "!H")
         return address, port
 
-    def tcp_loop(self, sock1, sock2):
+    def tcp_loop(self, csock, ssock):
+        global inbound_traffic
+        global outbound_traffic
         while True:
-            # wait until client or remote is available for read
-            r, _, _ = select([sock1, sock2], [], [], IDLE_TIMEOUT)
+            r, _, _ = select([csock, ssock], [], [], IDLE_TIMEOUT)
             if not r:
                 raise socket.timeout()
 
-            if sock1 in r:
-                data = sock1.recv(4096)
+            if csock in r:
+                data = csock.recv(4096)
                 if not data:
                     break
-                sock2.sendall(data)
+                ssock.sendall(data)
+                with traffic_lock:
+                    outbound_traffic += len(data)
 
-            if sock2 in r:
-                data = sock2.recv(4096)
+            if ssock in r:
+                data = ssock.recv(4096)
                 if not data:
                     break
-                sock1.sendall(data)
+                csock.sendall(data)
+                with traffic_lock:
+                    inbound_traffic += len(data)
+
 
     def handle_connect(self, address, port):
         log_tag = '%s:%s -> %s:%s' % (self.client_address + (address, port))
-        try:
-            remote = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-            if CONNECT_HOST:
-                remote.bind((CONNECT_HOST, 0))
-            remote.connect((address, port))
-            logging.info('%s: connected', log_tag)
-        except Exception as err:
-            logging.error('%s: connect error %s', log_tag, err)
-            self.send_reply(self.STATUS_ECONNREFUSED)
-            self.server.close_request(self.request)
-            return
 
-        self.send_reply(self.STATUS_SUCCEEDED)
+        if isinstance(address, dict):
+            ipv4 = address.get('ipv4')
+            ipv6 = address.get('ipv6')
+        else:
+            if ':' in address:
+                ipv6 = address
+                ipv4 = None
+            else:
+                ipv4 = address
+                ipv6 = None
+
+        if (ipv6 is not None and "ipv6" in CONNECT_HOST and CONNECT_HOST["ipv6"]) and \
```

Comment:
There are still redundant checks for `"ipv6" in CONNECT_HOST`, etc., and also now redundant checks for `if ipv6 is not None` in the `if` body below.

Instead of redundant `in` checks, maybe we should just split `CONNECT_HOST` into two separate variables, i.e. `CONNECT_HOST_IPV4` and `CONNECT_HOST_IPV6`. That will be more readable anyway, and there's rarely a need to package the two into a single data structure.